### PR TITLE
Lock wait_conditions to `oc|kubectl wait` pattern

### PR DESCRIPTION
**Pull request by cjeanner**
In order to avoid any misusage of the wait_conditions, and the
user-provided extension of those conditions, this patch ensures any
command passed for "wait condition" is a real "wait"

For instance, if anyone passes `cat /etc/passwd`, or `oc rsh` and
related, the assertion will fail, ensuring we don't have something that
should be either in a new stage, or in a post_stage_run hook.

An equivalent check is also existing in the architecture repository via
https://github.com/openstack-k8s-operators/architecture/pull/183

#### WARNING
This patch also remove the deprecated "validations" entry, since it has
been replaced by "wait_conditions", and architecture repository has a
check on the YAML structure

Co-Authored-By: @lewisdenny 

As a pull request owner and reviewers, I checked that:
- [X] Appropriate testing is done and actually running
(tested both situation, with a valid and an invalid pattern)

### CODE REVIEW

File: roles/kustomize_deploy/tasks/execute_step.yml
Comment (lewisdenny) [inline roles/kustomize_deploy/tasks/execute_step.yml:13]:
```suggestion
- name: Assert all wait conditions are valid
  vars:
    _stage_name: "stage_{{ stage_id }}"
    _custom_conditions: >-
      {{
        cifmw_architecture_wait_condition[_stage_name] |
        default([])
      }}
    _pattern: '^(oc|kubectl).*wait.*$'
  block:
    - name: Check custom wait conditions
      when: _custom_conditions | length > 0
      ansible.builtin.assert:
        that:
          - item is match(_pattern)
        msg: >-
          The following custom wait condition is not vaild: {{ item }}
        quiet: true
      ignore_errors: true
      register: custom_wait_conditions_results
      loop: "{{ _custom_conditions }}"

    - name: Check builtin wait conditions
      ansible.builtin.assert:
        that:
          - item is match(_pattern)
        msg: >-
          The following builtin wait condition is not vaild: {{ item }}
        quiet: true
      ignore_errors: true
      register: builtin_wait_conditions_results
      loop: "{{ stage.wait_conditions }}"

    - name: Check wait condition validation results
      ansible.builtin.fail:
        msg: "Please review the above failed wait conditions task"
      when: >
        (builtin_wait_conditions_results is failed or
        custom_wait_conditions_results is failed)
```

**Review by lewisdenny — CHANGES_REQUESTED** [review]
Hi Cédric, I tested this and think we can improve the output to be more helpful. 

I've left two comments, one I believe needs to be fixed if you want to continue your route.

The second is a suggest to improve the functionality in the following ways
- Shows what wait_condition has failed and what type it is.
- Allows all to be tested before failing to capture all issues in one run

**Comment by lewisdenny** [comment]
Test playbook:
```
---
- name: Test Playbook
  hosts: localhost
  gather_facts: false
  vars:
    stage_id: 1
    stage:
      wait_conditions:
        - >-
          oc -n openstack wait
          osdpd edpm-scale-out --for condition=Ready
          --timeout=1500s
        - >-
          cat other_file_to_check
    cifmw_architecture_wait_condition:
      stage_1:
        - >-
          oc -n openstack wait
          osdpd edpm-scale-out --for condition=Ready
          --timeout=1500s
        - >-
          cat file_to_check
  tasks:
    - name: Assert all wait conditions are valid
      vars:
        _stage_name: "stage_{{ stage_id }}"
        _custom_conditions: >-
          {{
            cifmw_architecture_wait_condition[_stage_name] |
            default([])
          }}
        _pattern: '^(oc|kubectl).*wait.*$'
      block:
        - name: Check custom wait conditions
          when: _custom_conditions | length > 0
          ansible.builtin.assert:
            that:
              - item is match(_pattern)
            msg: >-
              The following custom wait condition is not vaild: {{ item }} \n
            quiet: true
          ignore_errors: true
          register: custom_wait_conditions_results
          loop: "{{ _custom_conditions }}"

        - name: Check builtin wait conditions
          ansible.builtin.assert:
            that:
              - item is match(_pattern)
            msg: >-
              The following builtin wait condition is not vaild: {{ item }}\n
            quiet: true
          ignore_errors: true
          register: builtin_wait_conditions_results
          loop: "{{ stage.wait_conditions }}"

        - name: Check wait condition validation results
          ansible.builtin.fail:
            msg: "Please review the above failed wait conditions task"
          when: >
            (builtin_wait_conditions_results is failed or
            custom_wait_conditions_results is failed)
```

Results:
```
❯ ansible-playbook test.yml

PLAY [Test Playbook] ******************************************************************************************************************************************************************************

TASK [Check custom wait conditions] ***************************************************************************************************************************************************************
ok: [localhost] => (item=oc -n openstack wait osdpd edpm-scale-out --for condition=Ready --timeout=1500s)
failed: [localhost] (item=cat file_to_check) => changed=false 
  ansible_loop_var: item
  assertion: item is match(_pattern)
  evaluated_to: false
  item: cat file_to_check
  msg: 'The following custom wait condition is not vaild: cat file_to_check'
...ignoring

TASK [Check builtin wait conditions] **************************************************************************************************************************************************************
ok: [localhost] => (item=oc -n openstack wait osdpd edpm-scale-out --for condition=Ready --timeout=1500s)
failed: [localhost] (item=cat other_file_to_check) => changed=false 
  ansible_loop_var: item
  assertion: item is match(_pattern)
  evaluated_to: false
  item: cat other_file_to_check
  msg: 'The following builtin wait condition is not vaild: cat other_file_to_check'
...ignoring

TASK [Check wait condition validation results] ****************************************************************************************************************************************************
fatal: [localhost]: FAILED! => changed=false 
  msg: Please review the above failed wait conditions task

PLAY RECAP ****************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=2   
```

File: roles/kustomize_deploy/tasks/execute_step.yml
Comment (cjeanner) [reply roles/kustomize_deploy/tasks/execute_step.yml:17]: it comes from the upper loop:
https://github.com/openstack-k8s-operators/ci-framework/blob/6db3002cce631c1b85df6357be4cecd6d5bffedc/playbooks/06-deploy-architecture.yml#L127-L130

File: roles/kustomize_deploy/tasks/execute_step.yml
Comment (cjeanner) [reply roles/kustomize_deploy/tasks/execute_step.yml]: indeed, providing some more data in the error message is a good thing.

**Comment by cjeanner** [comment]
```
[...]
msg: 'The following custom condition is invalid: sudo cat /etc/passwd'
msg: 'The following custom condition is invalid: oc rsh exec blah'
msg: 'The following custom condition is invalid: kubectl rsh exec bar'
ok: [instance] => (item=oc -n openstack wait nncp -l osp/nncm-config-type=standard --for jsonpath='{.status.conditions[0].reason}'=SuccessfullyConfigured --timeout=60s)
[...]
TASK [kustomize_deploy : Check wait_conditions validation result] **************
fatal: [instance]: FAILED! => changed=false
  msg: Review and correct the faulty wait_conditions listed above.
```

**Review by marios — COMMENTED** [review]
looks sane but will wait for the review call before adding approve, to hear more about it ;)

File: roles/kustomize_deploy/tasks/execute_step.yml
Comment (cjeanner) [reply roles/kustomize_deploy/tasks/execute_step.yml:18]: (And why did I read "stage_id" in your question now... Yes, you're right, I missed that one, it has been added in the last iteration)

**Review by lewisdenny — APPROVED** [review]
Thanks Cédric, looks REALLY nice now!

**Pull request merged**
